cpp/src/arrow/util/concatenate.cc

```diff
@@ -0,0 +1,335 @@
+    auto get_values_range = [&](int in_i, UnionArray::type_id_t code) -> range& {
+      return values_ranges[in_i * in_size_ + code];
+    };
+    for (int i = 0; i != in_size_; ++i) {
+      auto codes = in_[i].buffers[1]->data();
+      auto offsets = reinterpret_cast<const int32_t*>(in_[i].buffers[2]->data());
+      for (auto index = offsets_[i]; index != offsets_[i] + lengths_[i]; ++index) {
+        get_values_range(i, codes[index]).widen_to_include(offsets[index]);
+      }
+    }
+
+    // for each type_code, use the min/max offset as a slice range
+    // and concatenate sliced data for that type_code
+    out_.child_data.resize(static_cast<int>(max_code) + 1);
+    for (auto code : u.type_codes()) {
+      std::vector<ArrayData> values_slices(in_size_);
+      for (int i = 0; i != in_size_; ++i) {
+        auto values_range = get_values_range(i, code);
+        values_slices[i] =
+            SliceData(*in_[i].child_data[code], values_range.offset, values_range.length);
+      }
+      auto values = std::make_shared<ArrayData>();
+      RETURN_NOT_OK(ConcatenateImpl(values_slices, pool_).Concatenate(values.get()));
+      out_.child_data[code] = values;
+    }
+
+    // for each input array, adjust the offsets by the length of the slice
+    // range for a given type code and the minimum offset in that input array,
+    // so that offsets point into the concatenated values
+    std::vector<int32_t> total_lengths(static_cast<int>(max_code) + 1, 0);
+    std::shared_ptr<Buffer> offsets_buffer;
+    RETURN_NOT_OK(AllocateBuffer(pool_, out_.length * sizeof(int32_t), &offsets_buffer));
+    auto raw_offsets = reinterpret_cast<int32_t*>(offsets_buffer->mutable_data());
+    for (int i = 0; i != in_size_; ++i) {
+      auto codes = in_[i].buffers[1]->data();
+      auto offsets = reinterpret_cast<const int32_t*>(in_[i].buffers[2]->data());
+      for (auto index = offsets_[i]; index != offsets_[i] + lengths_[i]; ++index) {
+        auto min_offset = get_values_range(i, codes[index]).offset;
+        *raw_offsets++ = offsets[index] - min_offset + total_lengths[codes[index]];
+      }
+      for (auto code : u.type_codes()) {
+        total_lengths[code] += get_values_range(i, code).length;
+      }
+    }
+    out_.buffers.push_back(offsets_buffer);
+    return Status::OK();
+    */
+  }
+
+  Status Visit(const ExtensionType&) {
+    // XXX can we just concatenate their storage?
+    return Status::NotImplemented("concatenation of extension arrays");
+  }
+
+  Status Concatenate(ArrayData* out) && {
+    std::shared_ptr<Buffer> null_bitmap;
+    if (out_.null_count != 0) {
+      RETURN_NOT_OK(ConcatenateBitmaps(0, &null_bitmap));
+    }
+    out_.buffers = {null_bitmap};
+    RETURN_NOT_OK(VisitTypeInline(*out_.type, this));
+    *out = std::move(out_);
+    return Status::OK();
+  }
+
+  Status ConcatenateBitmaps(int index, std::shared_ptr<Buffer>* bitmap_buffer) {
```

Comment:
I think I'd prefer seeing `std::vector<std::shared_ptr<Buffer>>` instead of `int index` even if it involves one more array and shared_ptr copies. This would mean you can extract ConcatenateBitmaps in bit-utils.h

Reply:
This would be slightly thorny: a bitmap is an offset and length as well as a buffer so the signature would be `Status ConcatenateBitmaps(std::vector<std::tuple<const uint8_t*, int64_t, int64_t>>&, std::shared_ptr<Buffer>*)`. I'll extract that tuple as `BitsView`

Reply:
Might be time to implement a `Bitmap` class :)

Reply:
https://issues.apache.org/jira/browse/ARROW-4722